compiler/src/org.graalvm.compiler.nodes/src/org/graalvm/compiler/nodes/memory/HeapAccess.java

```diff
@@ -34,17 +34,25 @@
      */
     enum BarrierType {
         /**
-         * Primitive stores which do not necessitate barriers.
+         * Primitive access which do not necessitate barriers.
          */
         NONE,
         /**
-         * Array object stores which necessitate precise barriers.
+         * Array object access.
          */
-        PRECISE,
+        ARRAY,
         /**
-         * Field object stores which necessitate imprecise barriers.
+         * Field object access.
          */
-        IMPRECISE
+        FIELD,
+        /**
+         * Unknown (aka field or array) object access.
+         */
+        UNKNOWN,
+        /**
+         * Referent field access.
+         */
+        REFERENT
```

Comment:
It's not ideal to have a barrier type that is only used by one specific VM, the HotSpot VM, in the VM-independent part of Graal. Maybe it is a bad idea in the first place to have barrier types as an `enum` here, and not just define an empty marker interface here and have the actual value definitions in VM-specific code.

Reply:
I think a clear set in the core would better than multiple different definitions but I agree this naming is very HotSpot specific.  Maybe this would be better called WEAK_FIELD to generalize it?

Reply:
We could do a split: some definitions (definitely `UNKNOWN`, but maybe also `FIELD` and `ARRAY`) in the platform independent part, and others like `REFERENT` in HotSpot specific code.

Reply:
Why is REFERENT HotSpot specific since it's a field in a (non-HotSpot) JDK class?

Reply:
I guess the field is defined in the `Reference` class because it simplified the HotSpot implementation (unfortunately the JDK and the VM are not as decoupled as they should be). But the field has the interesting comment `/* Treated specially by GC */` ;-)
So your argument is that the bad JDK/VM separation means that Graal also does not need to separate it out into VM independent code. Which is a valid argument.

Reply:
If SVM adds a G1 style collector it will need the exact same special handling.  The read from referent needs special handling because it's reading from a weak field.  Even if the field were hidden in some way the actual read would still need this work.

Reply:
Currently, only Shenandoah and G1, probably ZGC (not sure) need to know about it, they are all Hotspot-only. Current SVM GCs don't currently need to know about it, but it surely belongs into the GC interface anyway (even if ignored by specific implementations). I'm fine with renaming it to WEAK_FIELD, this should be generic enough?

Reply:
I think so.

Reply:
Latest push fixes this.